src/charm.py

```diff
@@ -206,6 +219,85 @@ def _on_peer_relation_changed(self, event: RelationChangedEvent) -> None:
         ):
             self.unit.status = ActiveStatus()
 
+    # flake8: noqa: C901
+    def _on_db_router_relation_changed(self, event: RelationChangedEvent) -> None:
+        """Handle the db_router relation changed event."""
+        if not self.unit.is_leader():
+            return
+
+        self.unit.status = MaintenanceStatus("Setting up db-router relation")
+
+        # Get the application data from the relation databag
+        # Abstracted is the fact that it is received from the leader unit of the related app
+        unit_address = str(self.model.get_binding(PEER).network.bind_address)
+        unit_names = " ".join([unit.name for unit in event.relation.units])
+
+        event_relation_databag = event.relation.data[self.unit]
+        event_relation_databag["db_host"] = json.dumps(unit_address)
+
+        application_unit = list(event.relation.units)[0]
+        application_data = copy.deepcopy(event.relation.data[application_unit])
+        application_info = collections.defaultdict(dict)
+
+        # Create the mysql router user if it does not already exist
+        try:
+            mysqlrouter_username = application_data.get("mysqlrouter_username")
+            mysqlrouter_user_exists = self._mysql.does_mysql_user_exist(mysqlrouter_username)
+
+            if mysqlrouter_username and not mysqlrouter_user_exists:
+                mysqlrouter_password = generate_random_password(PASSWORD_LENGTH)
+                self._mysql.configure_mysqlrouter_user(mysqlrouter_username, mysqlrouter_password)
+
+                event_relation_databag["mysqlrouter_password"] = json.dumps(mysqlrouter_password)
+
+            if mysqlrouter_username:
+                event_relation_databag["mysqlrouter_allowed_units"] = json.dumps(unit_names)
+        except (MySQLCheckUserExistenceError, MySQLConfigureRouterUserError):
+            self.unit.status = BlockedStatus("Failed to create mysqlrouter user")
+            return
+
+        # Retrieve and compile application usernames and database names
+        for key, value in application_data.items():
+            application_name = key.split("_")[0]
+
+            if application_name == "mysqlrouter":
+                continue
+
+            if "username" in key:
+                application_info[application_name]["username"] = value
+
+            if "database" in key:
+                application_info[application_name]["database"] = value
+
+        # Create the application database and scoped user if it does not already exist
+        for application_name, database_info in application_info.items():
```

Comment:
Is it possible to do this in one pass? Instead of populating application_info and then accessing from it, would it be logical to do something like the following?

```python
for key, value in application_data.items():
    application_name = key.split("_")[0]
    if application_name == "mysqlrouter":
        continue
    
    # Skip everything that's not a username so we aren't doing the below steps more than we have to
    if "username" not in key: 
        continue
    username = value
    database = application_data[f"{application_name}_database"]
    application_password = generate_random_password(PASSWORD_LENGTH)

    try: 
        if not self._mysql.does_mysql_user_exist(username):
            self._mysql.create_application_database_and_scoped_user(
                database,
                username,
                application_password,
            )
    # Carry on as normal from here
```

Reply:
good suggestion: refactored the method heavily to be less complex in 5a2d6a26ef82b30eb9937885ae5460ab5f22024f